phastSim/phastSim.py

```diff
@@ -641,23 +809,196 @@ def sampleMutation(self, allele, rates, rand):
+            node.belowNodes = [child0, child1]
+            node.rate = child0.rate + child1.rate
+            node.genomePos = [(self.n_insertions, pos_left), (self.n_insertions, pos_right)]
+            return node
+
+
+    def sampleInsertion(self, rate, node, rand, level):
+
+        # sample a geometric random variable
+        insertion_length = 1 + int(np.log(rand)/np.log(rate))
+        
+        # TODO generate a nucleotide sequence using the correct model
+        insertion = np.random.choice(self.allelesList, insertion_length, self.insertionFrequencies)
+
+        print("*************", insertion)
+        # turn the nucleotide sequence into a genomeTree
+        genomeSeq = "".join(insertion)
+
+        newNode = genomeNode(level=level)
+
+        # generate the genome tree of this insert from (node, genomeSeq)
+        subTree = self.buildTree(newNode, genomeSeq, 0, len(genomeSeq) - 1)
+
+        self.normalizeRates(subTree, self.indels, self.codon, len(genomeSeq), self.scale)
+
+        # build a bigger tree of the form t = (node, subTree). 
+        newRoot = genomeNode(level=level)
+        newRoot.belowNodes = [node, subTree]
+        newRoot.rate = node.rate + subTree.rate
+
+
+        mutData = mutation(mType=mType.INS, genomePos=node.refNode.genomePos[0], insert=genomeSeq, index=self.n_insertions)
+
+        return mutData, newRoot
 
     def findPos(self, rand, parentGenomeNode, level):
         # find position to mutate along the genome, and update temporary genome tree structure as you go
         if parentGenomeNode.isTerminal:
             # reached a terminal node, now sample the mutation event at the position and update all rates
             node = parentGenomeNode.refNode
             a = parentGenomeNode.allele
+            
+            # we'll choose an indel if rand is less than a threshold determined by the insertion and deletion rates for this site
+            if self.indels:
+
+                # deletions
+                if rand < node.deletionRate:
+
+                    attempted_deletion_length = self.sampleDeletion(node.deletionLength, rand/node.deletionRate)
+                    pos = node.genomePos[0]
+                    parentGenomeNode.rate = node.rate
+                    mutEvent = mutation(mType=mType.DEL, genomePos=pos, length=attempted_deletion_length)
+                    # in the tidy up function (self.deleteNodes), the length of this mutEvent may be changed if we are in an edge case
+                    # where we delete at a site very close to the end of the genome. The rates for all nodes are also updated there. 
+                    return mutEvent
+                
+                # else insertion:
+                if rand < node.deletionRate + node.insertionRate:
+                
+                    mutEvent, mutEventRootNode = self.sampleInsertion(node.insertionLength, node, rand/(node.deletionRate + node.insertionRate), level)
```

Comment:
I don't think this is correct: at this point rand will always be >deletionRate, and so it will not be a completely random number, meaning that you would end up always sampling larger insertion lengths.
Instead, it would be better to:

rand -= node.deletionRate
if rand < node.insertionRate:
     mutEvent, mutEventRootNode = self.sampleInsertion(node.insertionLength, node, rand/node.insertionRate, level)

Reply:
Also, on a completely separate issue, I think there is a problem with attaching the new inserted subtree to the existing genome tree layer. This is because the sampleInsertion() function is called on a terminal node "node", and then it creates a subtree which also has "node" as terminal, but it doesn't change the "belowNodes" links in "node"'s parent node, and so the current layer genome tree keeps the same shape, and the created genome subtree is basically ignored. I'm not sure what is the easiest fix. One possibility is that findPos() is given an additional argument that keeps track of the parent node of the currently considered node, and possibly also another argument that keeps track of if the current node is the first or second child of its parent. Then, these two arguments can be passed to sampleInsertion() which uses them to change the corresponding belowNodes link of "node"'s parent to be "newRoot" instead of "node". Does this make sense?

Reply:
You are right this is a bug, I'll fix it like you suggested

Reply:
I think the same thing also applies for substitutions too, so there needs to be rand -= node.insertionRate later on also